### Statistical Learning for Data Science 2 (229352)
#### Instructor: Donlapark Ponnoprat

#### [Course website](https://donlapark.pages.dev/229352/)

## Lab #4

 ต่างจากคราวที่แล้วคือใช้ Naive Bayes
 และใช้การ Randomized Search

In [1]:
from sklearn.datasets import fetch_20newsgroups
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics import classification_report
from sklearn.model_selection import GridSearchCV, RandomizedSearchCV

from sklearn.naive_bayes import MultinomialNB
from sklearn.pipeline import Pipeline

from scipy.stats import uniform

In [2]:
train = fetch_20newsgroups(subset='train')
test = fetch_20newsgroups(subset='test')

Xtrain = train.data[:3000]
ytrain = train.target[:3000]
Xtest = test.data[:500]
ytest = test.target[:500]

print("X:", len(Xtest))
print("y:", len(ytest))

X: 500
y: 500


### Naive Bayes [(Documentation)](https://scikit-learn.org/stable/modules/generated/sklearn.naive_bayes.MultinomialNB.html)

In [ ]:
tfidf = Tfidfvectorizer(stop_words='english')

Xtrain_tfidf = tfidf.fit_transform(Xtrain)
Xtest_tfidf = tfidf.transform(Xtest)

ืnb = MultinomialNB(alpha=2)

# หา Parameter ที่เหมาะสม
# เอา Parameter ที่ได้มา fit ใหม่ทั้งชุด

# เอา Model ที่ได้มา Predict

params = {'alpha': uniform(loc=0, scale=100)}

clf = RandomizedSearchCV(pipeline, params, n_iter=5)

clf.fit(Xtrain_tfidf, ytrain)

y_pred = clf.predict(Xtest)

### Random Search Cross-Validation [(Documentation)](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.RandomizedSearchCV.html)

### Uniform distribution in `Scipy` [(Documentation)](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.uniform.html)

In [ ]:
# params เราจะสุ่มค่า alpha จากการแจกแจงแบบไหน
params = {'alpha': uniform(loc=0, scale=100)} # ค่าต่ำสุด 0, ค่าสูงสุด 100 (สุ่มมาใน 0-100)

# จำนวนครั้งที่เราจะสุ่มมาทำ Grid Search แต่ละรอบคือ n_iter (Default = 10)
clf = RandomizedSearchCV(pipeline, params, n_iter=5)

clf.fit(Xtrain_tfidf, ytrain)

y_pred = clf.predict(Xtest)

#### Exercise

1. Create a **pipeline** for tf-idf and the Naive Bayes model, use grid search 5-fold cross-validation across different values of `alpha` to find the best model.

2. For the best value of `alpha`, compute the `f1_macro` score on the test set.
* What value of `alpha` did you obtain?
* What is the model's `f1_macro` score?

3. Repeat Exercise 1 and 2 for **random search** 5-fold cross validation across different values of `alpha`. Compute the `f1_macro` score on the test set.
* What value of `alpha` did you obtain?
* Did you get a better `f1_macro` score compared to grid search in Exercise 2?

In [4]:
# กำหนดตัวแปร tfidf
tfidf = TfidfVectorizer(stop_words="english")
Xtrain_tfidf = tfidf.fit_transform(Xtrain)
Xtest_tfidf = tfidf.transform(Xtest)
Xtrain_tfidf

<3000x61689 sparse matrix of type '<class 'numpy.float64'>'
	with 341881 stored elements in Compressed Sparse Row format>

In [6]:
# สร้าง pipeline สำหรับทำ Grid Search
pipeline = Pipeline([('tfidf', TfidfVectorizer()),
                     ('nb', MultinomialNB())])
pipeline.fit(Xtrain, ytrain)
ypred = pipeline.predict(Xtest)

print(classification_report(ytest, ypred))

              precision    recall  f1-score   support

           0       0.83      0.24      0.37        21
           1       0.83      0.24      0.37        21
           2       0.71      0.65      0.68        26
           3       0.73      0.56      0.63        34
           4       0.73      0.79      0.76        34
           5       0.95      0.81      0.88        26
           6       1.00      0.50      0.67        22
           7       0.67      1.00      0.80        28
           8       0.96      0.73      0.83        33
           9       0.91      0.84      0.87        25
          10       0.85      0.85      0.85        27
          11       0.58      0.95      0.72        20
          12       0.56      0.38      0.45        24
          13       0.81      0.57      0.67        23
          14       0.85      0.39      0.54        28
          15       0.27      1.00      0.42        29
          16       0.44      0.95      0.61        21
          17       1.00    

/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [14]:
# ทำ Grid Search
# parameters = {"ชื่อโมเดล__ชื่อพารามิเตอร์ที่เรากำหนดไว้": [พารามิเตอร์]}
params = {"tfidf__stop_words": ['english'],
          "nb__alpha": [0.1, 1, 10, 50, 100]}
gridcv = GridSearchCV(pipeline, params,
                      scoring='f1_macro', cv=5)
gridcv.fit(Xtrain, ytrain)

GridSearchCV(cv=5,
             estimator=Pipeline(steps=[('tfidf', TfidfVectorizer()),
                                       ('nb', MultinomialNB())]),
             param_grid={'nb__alpha': [0.1, 1, 10, 50, 100],
                         'tfidf__stop_words': ['english']},
             scoring='f1_macro')

In [15]:
gridcv.best_estimator_

Pipeline(steps=[('tfidf', TfidfVectorizer(stop_words='english')),
                ('nb', MultinomialNB(alpha=0.1))])

In [16]:
print(gridcv.best_params_)
print(gridcv.best_score_)

{'nb__alpha': 0.1, 'tfidf__stop_words': 'english'}
0.8201560377058564


In [17]:
# Make predictions on the test set
ypred = gridcv.predict(Xtest)

# Report classification scores
print(classification_report(ytest, ypred))

              precision    recall  f1-score   support

           0       0.73      0.52      0.61        21
           1       0.68      0.62      0.65        21
           2       0.58      0.58      0.58        26
           3       0.70      0.68      0.69        34
           4       0.88      0.82      0.85        34
           5       0.89      0.62      0.73        26
           6       0.89      0.77      0.83        22
           7       0.79      0.96      0.87        28
           8       0.97      0.85      0.90        33
           9       0.92      0.88      0.90        25
          10       0.87      1.00      0.93        27
          11       0.77      1.00      0.87        20
          12       0.61      0.71      0.65        24
          13       0.83      0.83      0.83        23
          14       0.83      0.89      0.86        28
          15       0.60      0.93      0.73        29
          16       0.50      0.95      0.66        21
          17       0.94    

In [19]:
# สร้าง pipeline ใหม่เพื่อทำ Random Search
pipeline = Pipeline([('nb', MultinomialNB())])

random_params = {"nb__alpha": uniform(loc=0, scale=100)}
clf = RandomizedSearchCV(pipeline, random_params, n_iter=5,
                         scoring='f1_macro', cv=5, error_score='raise')
clf.fit(Xtrain_tfidf, ytrain)

RandomizedSearchCV(cv=5, error_score='raise',
                   estimator=Pipeline(steps=[('nb', MultinomialNB())]),
                   n_iter=5,
                   param_distributions={'nb__alpha': <scipy.stats._distn_infrastructure.rv_continuous_frozen object at 0x78f5c71d6ce0>},
                   scoring='f1_macro')

In [24]:
print(clf.best_params_)
print(clf.best_score_)

{'nb__alpha': 1.6780282655071144}
0.7445344662805201


In [25]:
clf.best_estimator_

Pipeline(steps=[('nb', MultinomialNB(alpha=1.6780282655071144))])

In [26]:
# Make predictions on the test set
ypred = clf.predict(Xtest_tfidf)

# Report classification scores
print(classification_report(ytest, ypred))

              precision    recall  f1-score   support

           0       0.73      0.38      0.50        21
           1       0.91      0.48      0.62        21
           2       0.56      0.73      0.63        26
           3       0.72      0.62      0.67        34
           4       0.66      0.85      0.74        34
           5       0.88      0.85      0.86        26
           6       1.00      0.64      0.78        22
           7       0.60      1.00      0.75        28
           8       0.90      0.82      0.86        33
           9       0.88      0.84      0.86        25
          10       0.82      1.00      0.90        27
          11       0.86      0.95      0.90        20
          12       0.57      0.54      0.55        24
          13       0.74      0.74      0.74        23
          14       0.90      0.64      0.75        28
          15       0.51      0.90      0.65        29
          16       0.49      0.90      0.63        21
          17       0.93    

/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


F1-score ของการทำ Grid Search = 0.73 ที่ alpha = 0.1

F1-score ของการทำ Random Search = 0.68 ที่ alpha = 1.6780282655071144

ดังนั้น Grid Search มีค่า F1-score ที่ดีกว่าค่า F1-score ของ Random Search